v5 migration guide (#3115)
Shopify/polaris-react: documentation/guides/migrating-from-v4-to-v5.md

      ).then((dictionary) => dictionary && dictionary.default);
    },
  });

  return (
-    <AppProvider i18n={i18n.translations.reverse()}>{/* App content */}</AppProvider>
+    <AppProvider i18n={i18n.translations}>{/* App content */}</AppProvider>
  );
}
```

## Build output changes - general consumers

Our compiled styles file has moved. Imports must be updated:

```diff
- import '@shopify/polaris/styles.css';
+ import '@shopify/polaris/dist/styles.css';
```

Our minified styles file has been removed, and you should use the above file instead.

```diff
- import '@shopify/polaris/styles.min.css';
+ import '@shopify/polaris/dist/styles.css';
```

Our scss entrypoint has been removed, and you should use the above styles file instead. If you require our mixins/functions you should also import the our public-api entrypoint.

```diff
- @import 'path_to_node_modules/@shopify/polaris/styles';
+ @import 'path_to_node_modules/@shopify/polaris/dist/styles.css';
+ /* Optional - only needed if you use our mixins/functions */
+ @import 'path_to_node_modules/@shopify/polaris/dist/styles/public-api';

## Build output changes - @shopify/sewing-kit integrations only

> Apps using `@shopify/sewing-kit` - Shopify's in-house opinionated build toolkit - will need to update to at least v0.132.2 to make use of the reworked "esnext" build output in polaris v5.

Reviewer: I don't think "Shopify's in-house opinionated build toolkit" adds value here.

```suggestion
Apps using `@shopify/sewing-kit` will need to update to at least v0.132.2 to make use of the reworked "esnext" build output in polaris v5.
```

Author: I was trying to enforce the idea that this doesn't apply to partners, and thus they can skip over this section. By adding this little bit I was trying to preempt thoughts of "what's sewing-kit? does this apply to me?" that may occur we just refer to sewing-kit by name.

Do you think just referring to the package by name is enough for partners to know this doesn't apply to them, is there there some other way we can help them reach that "cool, this doesn't apply to me" realization quickly?

Reviewer: I think referring to the package is enough. Happy for other opinions though.

Reviewer 2: I'm with Ben on this one, unless you feel strongly about removing this part of the sentence 👍

In v4 this order was `[fallbackDictionary, preferredDictionary]`. This was unintuitive, and does not align with other translation tools used at Shopify.
In v5 this order is `[preferredDictionary, fallbackDictionary]`.

If you are using `react-i18n` to provide async translations you can now pass translations in the order returned by `react-i18n`:

```diff
import AppProvider from '@shopify/polaris';
// en.json is English. Replace with fr.json for French, etc
import translations from '@shopify/polaris/locales/en.json';
import {useI18n} from '@shopify/react-i18n';

function App() {
  const [i18n] = useI18n({
    id: 'Polaris',
    fallback: translations,
    translations(locale) {
      return import(
        /* webpackChunkName: "Polaris-i18n", webpackMode: "lazy-once" */ `@shopify/polaris/locales/${locale}.json`
      ).then((dictionary) => dictionary && dictionary.default);
    },
  });

  return (
-    <AppProvider i18n={i18n.translations.reverse()}>{/* App content */}</AppProvider>
+    <AppProvider i18n={i18n.translations}>{/* App content */}</AppProvider>
  );
}
```

## Build output changes - general consumers

> Our compiled styles file has moved. Imports must be updated:

Reviewer 2: ```suggestion
The compiled styles file has moved. Imports must be updated:
```

```diff
- import '@shopify/polaris/styles.css';
+ import '@shopify/polaris/dist/styles.css';
```

> Our minified styles file has been removed, and you should use the above file instead.

Reviewer 2: ```suggestion
The minified styles file has been removed, and you should use the above file instead.
```

- import '@shopify/polaris/styles.min.css';
+ import '@shopify/polaris/dist/styles.css';
```

Our scss entrypoint has been removed, and you should use the above styles file instead. If you require our mixins/functions you should also import the our public-api entrypoint.

```diff
- @import 'path_to_node_modules/@shopify/polaris/styles';
+ @import 'path_to_node_modules/@shopify/polaris/dist/styles.css';
+ /* Optional - only needed if you use our mixins/functions */
+ @import 'path_to_node_modules/@shopify/polaris/dist/styles/public-api';
```

## Build output changes - @shopify/sewing-kit integrations only

Apps using `@shopify/sewing-kit` - Shopify's in-house opinionated build toolkit - will need to update to at least v0.132.2 to make use of the reworked "esnext" build output in polaris v5.

Apps using `@shopify/sewing-kit` and Polaris v4 also had to have an explicit import of `@shopify/polaris/esnext/styles/global.scss` in the same file as they imported and used `AppProvider`. This is no longer required, this import can be removed. The styles in this file will be implicitly imported when you use `AppProvider`.

```diff
- import '@shopify/polaris/esnext/styles/global.scss';
```

> Sewing-kit apps can use our sass public API by configuring their sass plugin's autoInclude option. The path to the sass API file has now changed and should be updated accordingly:

Reviewer 2: ```suggestion
Sewing-kit apps can use Polaris React’s Sass public API by configuring their Sass plugin's `autoInclude` option. The path to the Sass API file has now changed and should be updated accordingly:
```

```diff
plugins.sass({
  autoInclude: [
-    path.join(__dirname, 'node_modules/@shopify/polaris/esnext/styles/_public-api.scss'),
+    path.join(__dirname, 'node_modules/@shopify/polaris/dist/styles/_public-api.scss'),
  ],
 })
```

## Component API Changes

### ResourceList.FilterControl

The `ResourceList.FilterControl` subcomponent has been removed. You should use the `Filters` component to provide filtering options.

### EmptyState

The `EmptyState` component's `centeredLayout` prop has been removed as the illustration's location is now handled automatically by opting into the new design language. You should remove any usage of this prop.

### Removed Exports

#### NavigationMessageProps

> The `NavigationMessageProps` type has been removed, as Navigation's Message internal subcomponent has been removed.

Reviewer 2: ```suggestion
The `NavigationMessageProps` type has been removed, as `Navigation`'s `Message` internal subcomponent has been removed.
```

## Dependencies

> The peer dependencies on `react` and `react-dom` have been increased to 16.9.0, to allow us to use react's Profiler component in the future. Use the package manager of your choice to install a recent version of these pacakges.

Reviewer 2: For consistency

```suggestion
The peer dependencies on `react` and `react-dom` have been increased to `16.9.0`, opening the door to using the [React Profiler](https://reactjs.org/blog/2018/09/10/introducing-the-react-profiler.html) in the future. Use the package manager of your choice to install a recent version of these packages.
```

Apps built using `@shopify/sewing-kit` - Shopify's in-house opinionated build toolkit - will need to upgrade to at least version `0.132.2`.

## CDN usage

> As a result of build output changes, the location of our static stylesheet for CSS-only usage has been moved. Update any stylesheet imports to use the new path:

Reviewer 2: I try to avoid "our", in case it's perceived as "not yours".

```suggestion
As a result of build output changes, the location of the static stylesheet for CSS-only usage has been moved. Update any stylesheet imports to use the new path:
```

import translations from '@shopify/polaris/locales/en.json';
import {useI18n} from '@shopify/react-i18n';

function App() {
  const [i18n] = useI18n({
    id: 'Polaris',
    fallback: translations,
    translations(locale) {
      return import(
        /* webpackChunkName: "Polaris-i18n", webpackMode: "lazy-once" */ `@shopify/polaris/locales/${locale}.json`
      ).then((dictionary) => dictionary && dictionary.default);
    },
  });

  return (
-    <AppProvider i18n={i18n.translations.reverse()}>{/* App content */}</AppProvider>
+    <AppProvider i18n={i18n.translations}>{/* App content */}</AppProvider>
  );
}
```

## Build output changes - general consumers

Our compiled styles file has moved. Imports must be updated:

```diff
- import '@shopify/polaris/styles.css';
+ import '@shopify/polaris/dist/styles.css';
```

Our minified styles file has been removed, and you should use the above file instead.

```diff
- import '@shopify/polaris/styles.min.css';
+ import '@shopify/polaris/dist/styles.css';
```

Our scss entrypoint has been removed, and you should use the above styles file instead. If you require our mixins/functions you should also import the our public-api entrypoint.

> ```diff
> 
> - @import 'path_to_node_modules/@shopify/polaris/styles';
> + @import 'path_to_node_modules/@shopify/polaris/dist/styles.css';
> + /* Optional - only needed if you use our mixins/functions */
> + @import 'path_to_node_modules/@shopify/polaris/dist/styles/public-api';
> ```

Reviewer: ```suggestion
```diff
- @import 'path_to_node_modules/@shopify/polaris/styles';
+ @import 'path_to_node_modules/@shopify/polaris/dist/styles.css';
+ /* Optional - only needed if you use our mixins/functions */
+ @import 'path_to_node_modules/@shopify/polaris/dist/styles/public-api';
```
```

Author: Pretty sure you didn't want want to get rid of that final ``` that closes the  fenced code block - that might be nested codeblocks in markdown being an ass I think that's bit me before. I've added it back :)

# Migrating from v4 to v5

Polaris v5.0.0 ([full release notes](https://github.com/Shopify/polaris-react/releases/tag/v5.0.0)) removes most features deprecated in v4.x.x releases. This file describes all code updates required to stay up to date.

## AppBridge integrations removed

> In v3 and v4 `AppProvider` could accept `apiKey` `shopOrigin` and `forceRedirect` props that would configure AppBridge delegation in our `ResourcePicker`, `Loading`, `Modal`, `Page` and `Toast` components.

Reviewer: > In v3 and v4 `AppProvider` could accept `apiKey` `shopOrigin` and `forceRedirect` props that would configure AppBridge delegation in our `ResourcePicker`, `Loading`, `Modal`, `Page` and `Toast` components.

I think we should rephrase this to say that in v5 these props are removed. I know the title above says integrations removed but it's not clear if these are the integrations that are removed or what works in v3 and v4.

The `ResourcePicker` component has been fully removed is it only had meaning within an AppBridge app. Other components retain their non-app-bridge behaviour.

AppBridge integration has been removed and abstracted into a separate `@shopify/app-bridge-react` package. If you require the AppBridge implementation of these components then use the components from the [`@shopify/app-bridge-react` package](https://shopify.dev/tools/app-bridge/react-components).

```js
// Old
import {AppProvider, Modal} from '@shopify/polaris';

function MyApp() {
  const config = {apiKey: '12345', shopOrigin: shopOrigin};

  return (
    <AppProvider apiKey={config.apiKey} shopOrigin={config.shopOrigin}>
      <Modal /* ... *//>
    </AppProvider>
  );
}

> // NEw

Reviewer: ```suggestion
// New
```

import {AppProvider} from '@shopify/polaris';
import {Provider, TitleBar} from '@shopify/app-bridge-react';

function MyApp() {
  const config = {apiKey: '12345', shopOrigin: shopOrigin};

  return (
    <AppProvider>
      <Provider config={config}>
        <Modal /* ... *//>
      </Provider>
    <AppProvider>
  );
}
```

## AppProvider's i18n translations array is now preferred language first

> AppProvider's i18n prop accepts an array of translation dictionaries so that if
> a translation key in one language is not found it can try a fallback language.

Reviewer: ```suggestion
AppProvider's i18n prop accepts an array of translation dictionaries so that if a translation key in one language is not found it can try a fallback language.
```

# Migrating from v4 to v5

Polaris v5.0.0 ([full release notes](https://github.com/Shopify/polaris-react/releases/tag/v5.0.0)) removes most features deprecated in v4.x.x releases. This file describes all code updates required to stay up to date.

## AppBridge integrations removed

In v3 and v4 `AppProvider` could accept `apiKey` `shopOrigin` and `forceRedirect` props that would configure AppBridge delegation in our `ResourcePicker`, `Loading`, `Modal`, `Page` and `Toast` components. This functionality has been removed.

In v5 AppBridge integration has been removed from these components:

- The `AppProvider` component's `apiKey`, `shopOrigin` and `forceRedirect` have been removed.
- The `Modal` component's `size` and `message` props have been removed as they only had meaning within an AppBridge app.
- The `Page` component's `forceRender` prop has been removed as it only had meaning within an AppBridge app.
- The `ResourcePicker` component has been removed as it only had meaning within an AppBridge app.

> AppBridge behavour has been abstracted into a separate `@shopify/app-bridge-react` package. If you require the AppBridge implementation of these components then use the components from the [`@shopify/app-bridge-react` package](https://shopify.dev/tools/app-bridge/react-components).

Reviewer 3: ```suggestion
AppBridge behavior has been abstracted into a separate `@shopify/app-bridge-react` package. If you require the AppBridge implementation of these components then use the components from the [`@shopify/app-bridge-react` package](https://shopify.dev/tools/app-bridge/react-components).
```

```js
// Old
import {AppProvider, Modal} from '@shopify/polaris';

function MyApp() {
  const config = {apiKey: '12345', shopOrigin: shopOrigin};

  return (
    <AppProvider apiKey={config.apiKey} shopOrigin={config.shopOrigin}>
      <Modal /* ... *//>
    </AppProvider>
  );
}

// New
import {AppProvider} from '@shopify/polaris';

> import {Provider, TitleBar} from '@shopify/app-bridge-react';

Reviewer 4: ```suggestion
import {Provider, Modal} from '@shopify/app-bridge-react';
```

function MyApp() {
  const config = {apiKey: '12345', shopOrigin: shopOrigin};

  return (
    <AppProvider>
      <Provider config={config}>
        <Modal /* ... *//>
      </Provider>
    <AppProvider>
  );
}
```

> ## AppProvider's i18n translations array is now preferred language first

Reviewer 4: ```suggestion
## AppProvider's i18n translations array is now the preferred language first
```

> ## AppProvider's i18n translations array is now preferred language first

Author: ```suggestion
## AppProvider's i18n translations array is now ordered by preferred language first
```

AppProvider's i18n prop accepts an array of translation dictionaries so that if a translation key in one language is not found it can try a fallback language.

In v4 this order was `[fallbackDictionary, preferredDictionary]`. This was unintuitive, and does not align with other translation tools used at Shopify.

In v5 this order is `[preferredDictionary, fallbackDictionary]`. You should reverse the order of the array you pass into the i18n prop.

```diff
import enTranslations from '@shopify/polaris/locales/en.json';
import frTranslations from '@shopify/polaris/locales/fr.json';

function App() {
  return (
    // French translations as the primary language, english as a fallback
-    <AppProvider i18n={[enTranslations, frTranslations]}>

> +    <AppProvider i18n={frTranslations, enTranslations]}>

Reviewer 2: ```suggestion
+    <AppProvider i18n={[frTranslations, enTranslations]}>
```

```diff
import AppProvider from '@shopify/polaris';
// en.json is English. Replace with fr.json for French, etc
import translations from '@shopify/polaris/locales/en.json';
import {useI18n} from '@shopify/react-i18n';

function App() {
  const [i18n] = useI18n({
    id: 'Polaris',
    fallback: translations,
    translations(locale) {
      return import(
        /* webpackChunkName: "Polaris-i18n", webpackMode: "lazy-once" */ `@shopify/polaris/locales/${locale}.json`
      ).then((dictionary) => dictionary && dictionary.default);
    },
  });

  return (
-    <AppProvider i18n={i18n.translations.reverse()}>{/* App content */}</AppProvider>
+    <AppProvider i18n={i18n.translations}>{/* App content */}</AppProvider>
  );
}
```

## Build output changes - general consumers

Our compiled styles file has moved. Imports must be updated:

```diff
- import '@shopify/polaris/styles.css';
+ import '@shopify/polaris/dist/styles.css';
```

Our minified styles file has been removed, and you should use the above file instead.

```diff
- import '@shopify/polaris/styles.min.css';
+ import '@shopify/polaris/dist/styles.css';
```

> Our scss entrypoint has been removed, and you should use the above styles file instead. If you require our mixins/functions you should also import the our public-api entrypoint.

Reviewer 2: ```suggestion
The scss entrypoint has been removed, and you should use the above `styles.css` file instead. If you require Polaris React’s Sass mixins and functions, you should also import the `public-api` entrypoint.
```

> The scss entrypoint has been removed, and you should use the above `styles.css` file instead. If you require Polaris React’s Sass mixins and functions, you should also import the `public-api` entrypoint.

Reviewer 4: ```suggestion
The scss entry point has been removed, and you should use the above `styles.css` file instead. If you require Polaris React’s Sass mixins and functions, you should also import the `public-api` entry point.
```

```diff
- @import 'path_to_node_modules/@shopify/polaris/styles';
+ @import 'path_to_node_modules/@shopify/polaris/dist/styles.css';
+ /* Optional - only needed if you use our mixins/functions */
+ @import 'path_to_node_modules/@shopify/polaris/dist/styles/public-api';
```

## Build output changes - @shopify/sewing-kit integrations only

> Apps built using `@shopify/sewing-kit` - Shopify's in-house opinionated build toolkit - will need to update to at least v0.132.2 to make use of the reworked "esnext" build output in polaris v5.

Reviewer 2: ```suggestion
Apps built using `@shopify/sewing-kit` - Shopify's in-house opinionated build toolkit - will need to update to at least v0.132.2 to make use of the reworked "esnext" build output in Polaris React v5.
```

```diff
- import '@shopify/polaris/esnext/styles/global.scss';
```

Sewing-kit apps can use Polaris React’s Sass public API by configuring their Sass plugin's `autoInclude` option. The path to the Sass API file has now changed and should be updated accordingly:

```diff
plugins.sass({
  autoInclude: [
-    path.join(__dirname, 'node_modules/@shopify/polaris/esnext/styles/_public-api.scss'),
+    path.join(__dirname, 'node_modules/@shopify/polaris/dist/styles/_public-api.scss'),
  ],
 })
```

## Component API Changes

### ResourceList.FilterControl

The `ResourceList.FilterControl` subcomponent has been removed. You should use the `Filters` component to provide filtering options.

### Page

The `Page` component's `singleColumn` prop has been removed. You should replace any usage of this prop with the `narrowWidth` prop which has identical functionality.

### EmptyState

> The `EmptyState` component's `centeredLayout` prop has been removed as the illustration's location is now handled automatically by opting into the new design language. You should remove any usage of this prop.

Reviewer 4: ```suggestion
The `EmptyState` component's `centeredLayout` prop has been removed as the illustration's location is now handled automatically by opting in to the new design language. You should remove any usage of this prop.
```

# Migrating from v4 to v5

Polaris v5.0.0 ([full release notes](https://github.com/Shopify/polaris-react/releases/tag/v5.0.0)) removes most features deprecated in v4.x.x releases. This file describes all code updates required to stay up to date.

## AppBridge integrations removed

In v3 and v4 `AppProvider` could accept `apiKey` `shopOrigin` and `forceRedirect` props that would configure AppBridge delegation in our `ResourcePicker`, `Loading`, `Modal`, `Page` and `Toast` components. This functionality has been removed.

In v5 AppBridge integration has been removed from these components:

> - The `AppProvider` component's `apiKey`, `shopOrigin` and `forceRedirect` have been removed.

Reviewer 5: ```suggestion
- The `AppProvider` component’s `apiKey`, `shopOrigin` and `forceRedirect` have been removed.
```
I'm not sure this is important in markdown files but I try to use curly quotes for public facing docs
